runtime/doc/api.txt

```diff
@@ -659,13 +659,32 @@ nvim_echo({chunks}, {history}, {opts})                           *nvim_echo()*
                    (optional) name or ID `hl_group`.
       • {history}  (`boolean`) if true, add to |message-history|.
       • {opts}     (`vim.api.keyset.echo_opts`) Optional parameters.
+                   • id: message id for updating existing message.
                    • err: Treat the message like `:echoerr`. Sets `hl_group`
                      to |hl-ErrorMsg| by default.
                    • kind: Set the |ui-messages| kind with which this message
                      will be emitted.
                    • verbose: Message is controlled by the 'verbose' option.
                      Nvim invoked with `-V3log` will write the message to the
                      "log" file instead of standard output.
+                   • title: The title for |progress-message|.
+                   • status: Current status of the |progress-message|. Can be
+                     one of the following values
+                     • success: The progress item completed successfully
+                     • running: The progress is ongoing
+                     • failed: The progress item failed
+                     • cancel: The progressing process should be canceled.
+                       note: Cancel needs to be handled by progress initiator
+                       by listening for the `Progress` event
+                   • percent: How much progress is done on the progress
+                     message
+                   • data: dictionary containing additional information
+
+    Return: ~
+        (`integer`) Message id. Valid message id is always greater then 0
+        • -1 means nvim_echo didn't show a message
+        • 0 means nvim_echo didn't allocate a message id for the message.
+          happens for temp messages not stored in message history.
```

Comment:
Linking the message ID to whether or not a message is stored in the history seems wrong to me. There is no reason a temporary message shouldn't be able to be replaced in the UI.

Reply:
If we give IDs to temp messages, then we can't really validate the ids. We'll just have to trust the user and forward it.